gt/main.c

```diff
@@ -962,14 +958,21 @@ prep_notify_pkt(struct ggu_notify_pkt *ggu_pkt, struct gatekeeper_if *iface)
 		struct rte_ipv6_hdr *notify_ipv6 =
 			rte_pktmbuf_mtod_offset(ggu_pkt->buf,
 				struct rte_ipv6_hdr *, ggu_pkt->buf->l2_len);
+		/*
+		 * Distinct from @dgram_len_be because the IPv6
+		 * payload field could in theory include the length
+		 * of any extension headers.
+		 */
 		notify_ipv6->payload_len = rte_cpu_to_be_16(
 			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len -
 			sizeof(struct rte_ipv6_hdr));
 
-		notify_udp = (struct rte_udp_hdr *)&notify_ipv6[1];
+		notify_udp = rte_pktmbuf_mtod_offset(ggu_pkt->buf,
+			struct rte_udp_hdr *, ggu_pkt->buf->l3_len);
```

Comment:
Don't we have to add `ggu_pkt->buf->l2_len` as well here? The same comment applies to the IPv4 case.